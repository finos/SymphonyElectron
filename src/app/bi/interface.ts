export enum SDACrashProcess {
  MAIN = 'main',
  RENDERER = 'renderer',
  GPU = 'gpu',
}

export interface ICrashData extends IAnalyticsData {
  process: SDACrashProcess;
  crashCause: string;
  windowName: string;
  miniDump?: string;
}

export interface IInstallData extends IAnalyticsData {
  extra_data?: {
    installLocation: string;
    installType: string;
  };
}

export enum MenuActionTypes {
  AUTO_LAUNCH_ON_START_UP = 'auto_launch_on_start_up',
  ALWAYS_ON_TOP = 'always_on_top',
  MINIMIZE_ON_CLOSE = 'minimize_on_close',
  FLASH_NOTIFICATION_IN_TASK_BAR = 'flash_notification_in_task_bar',
  HAMBURGER_MENU = 'hamburger_menu',
  REFRESH_APP_IN_IDLE = 'refresh_app_in_idle',
}

export enum ScreenSnippetActionTypes {
  SCREENSHOT_TAKEN = 'screenshot_taken',
  ANNOTATE_ADDED_PEN = 'annotate_added_pen',
  ANNOTATE_ADDED_HIGHLIGHT = 'annotate_added_highlight',
  ANNOTATE_ADD = 'annotate_done',
  ANNOTATE_CLEARED = 'annotate_cleared',
  ANNOTATE_ERASED = 'annotate_erased',
  ANNOTATE_COPY = 'annotate_copy',
  ANNOTATE_SAVE_AS = 'annotate_save_as',
  ANNOTATE_CLOSE = 'annotate_close',
}

export enum ToastNotificationActionTypes {
  TOAST_CLOSED = 'toast_closed',
}

export enum SDAUserSessionActionTypes {
  Start = 'Start',
  End = 'End',
  Login = 'Login',
  Logout = 'Logout',
  Crash = 'Crash',
  ForceReload = 'Force_reload',
}

export enum InstallActionTypes {
  InstallStarted = 'Install_started',
  InstallCompleted = 'Install_completed',
  InstallFailed = 'Install_failed',
}

export enum InstallTypes {
  Auto = 'auto',
  Manual = 'manual',
}

export enum InstallLocationTypes {
  PROG_FILES = 'PROG_FILES',
  REMOTE = 'REMOTE',
  LOCAL = 'LOCAL',
  CUSTOM = 'CUSTOM',
}

export enum SDAEndReasonTypes {
  Reboot = 'Reboot',
  Closed = 'Closed',
  Crashed = 'Crashed',
}

export enum AnalyticsActions {
  ENABLED = 'ON',
  DISABLED = 'OFF',
}

export enum AnalyticsElements {
  MENU = 'Menu',
  SCREEN_CAPTURE_ANNOTATE = 'screen_capture_annotate',
  TOAST_NOTIFICATION = 'toast_notification',
  SDA_CRASH = 'sda_crash',
  SDA_SESSION = 'sda_session',
  SDA_INSTALL = 'sda_install',
}

export interface IAnalyticsData {
  element: AnalyticsElements;
  action_type?:
    | MenuActionTypes
    | ScreenSnippetActionTypes
    | ToastNotificationActionTypes
    | SDAUserSessionActionTypes
    | InstallActionTypes;
  action_result?: AnalyticsActions;
  extra_data?: object;
}

export interface ISessionData extends IAnalyticsData {
  extra_data?: {
    sessionStartDatetime?: string;
    machineStartDatetime?: string;
    machineId?: string;
    InstallVariant?: string;
    osName?: string;
    osVersion?: string;
    osLanguage?: string;
    osTimeZone?: string;
    cpuNumberOfCores?: number;
    cpuMaxFrequency?: number;
    cpuUsagePercent?: number;
    maxCPUUsagePercent?: number;
    memoryTotal?: number;
    memoryUsedPercent?: number;
    maxMemoryUsedPercent?: number;
    sdaUsedMemory?: number;
    memoryAvailable?: number;
    vdi?: boolean;
    endReason?: string;
    crashProcess?: string;
  };
}
