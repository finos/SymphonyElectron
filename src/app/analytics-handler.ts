import { WebContents } from 'electron';

export interface IAnalyticsData {
  element: AnalyticsElements;
  action_type?:
    | MenuActionTypes
    | ScreenSnippetActionTypes
    | ToastNotificationActionTypes;
  action_result?: AnalyticsActions;
  extra_data?: object;
}

export interface ICrashData extends IAnalyticsData {
  process: SDACrashProcess;
  crashCause: string;
  windowName: string;
  miniDump?: string;
}

export enum MenuActionTypes {
  AUTO_LAUNCH_ON_START_UP = 'auto_launch_on_start_up',
  ALWAYS_ON_TOP = 'always_on_top',
  MINIMIZE_ON_CLOSE = 'minimize_on_close',
  FLASH_NOTIFICATION_IN_TASK_BAR = 'flash_notification_in_task_bar',
  HAMBURGER_MENU = 'hamburger_menu',
  REFRESH_APP_IN_IDLE = 'refresh_app_in_idle',
}

export enum ScreenSnippetActionTypes {
  SCREENSHOT_TAKEN = 'screenshot_taken',
  ANNOTATE_ADDED_PEN = 'annotate_added_pen',
  ANNOTATE_ADDED_HIGHLIGHT = 'annotate_added_highlight',
  ANNOTATE_DONE = 'annotate_done',
  ANNOTATE_CLEARED = 'annotate_cleared',
  ANNOTATE_ERASED = 'annotate_erased',
}

export enum ToastNotificationActionTypes {
  TOAST_CLOSED = 'toast_closed',
}

export enum AnalyticsActions {
  ENABLED = 'ON',
  DISABLED = 'OFF',
}

export enum AnalyticsElements {
  MENU = 'Menu',
  SCREEN_CAPTURE_ANNOTATE = 'screen_capture_annotate',
  TOAST_NOTIFICATION = 'toast_notification',
  SDA_CRASH = 'sda_crash',
}

export enum SDACrashProcess {
  MAIN = 'main',
  RENDERER = 'renderer',
  GPU = 'gpu',
}

const MAX_EVENT_QUEUE_LENGTH = 50;
const analyticsCallback = 'analytics-callback';

class Analytics {
  private preloadWindow: WebContents | undefined;
  private analyticsEventQueue: IAnalyticsData[] = [];

  /**
   * Stores the reference to the preload window
   *
   * @param webContents {WeContents}
   */
  public registerPreloadWindow(webContents: WebContents): void {
    this.preloadWindow = webContents;

    if (!(this.preloadWindow && !this.preloadWindow.isDestroyed())) {
      return;
    }
    if (this.analyticsEventQueue && this.analyticsEventQueue.length > 0) {
      this.analyticsEventQueue.forEach((eventData) => {
        if (this.preloadWindow && !this.preloadWindow.isDestroyed()) {
          if (eventData.element === AnalyticsElements.SDA_CRASH) {
            eventData = eventData as ICrashData;
          }
          this.preloadWindow.send(analyticsCallback, eventData);
        }
      });
      this.resetAnalytics();
    }
  }

  /**
   * Sends the analytics events to the web client
   *
   * @param eventData {IAnalyticsData}
   */
  public track(eventData: IAnalyticsData): void {
    if (eventData.element === AnalyticsElements.SDA_CRASH) {
      eventData = eventData as ICrashData;
    }
    if (this.preloadWindow && !this.preloadWindow.isDestroyed()) {
      this.preloadWindow.send(analyticsCallback, eventData);
      return;
    }
    this.analyticsEventQueue.push(eventData);
    // don't store more than specified limit. keep most recent events.
    if (this.analyticsEventQueue.length > MAX_EVENT_QUEUE_LENGTH) {
      this.analyticsEventQueue.shift();
    }
  }

  /**
   * Clears the analytics queue
   */
  public resetAnalytics(): void {
    this.analyticsEventQueue = [];
  }
}

const analytics = new Analytics();

export { analytics };
