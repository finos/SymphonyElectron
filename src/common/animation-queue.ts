import { logger } from './logger';

export class AnimationQueue {
  private queue: any[] = [];
  private running: boolean = false;

  constructor() {
    this.animate = this.animate.bind(this);
  }

  /**
   * Pushes each animation to a queue
   *
   * @param object
   */
  public push(object): void {
    if (this.running) {
      this.queue.push(object);
      return;
    }
    this.running = true;
    setTimeout(() => this.animate(object), 0);
  }

  /**
   * Animates an animation that is part of the queue
   * @param object
   */
  public async animate(object): Promise<void> {
    try {
      await object.func.apply(null, object.args);
    } catch (err: any) {
      logger.error(
        `animation-queue: encountered an error: ${err} with stack trace: ${err.stack}`,
      );
    } finally {
      if (this.queue.length > 0) {
        // Run next animation
        this.animate.call(this, this.queue.shift());
      } else {
        this.running = false;
      }
    }
  }

  /**
   * Clears the queue
   */
  public clear(): void {
    this.queue = [];
  }
}
