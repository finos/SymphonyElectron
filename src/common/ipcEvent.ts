export enum ScreenShotAnnotation {
  COPY_TO_CLIPBOARD = 'copy-to-clipboard',
  SAVE_AS = 'save-as',
  CLOSE = 'close-snippet',
  UPLOAD = 'upload-snippet',
}
