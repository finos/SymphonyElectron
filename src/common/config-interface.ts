import { IConfig } from '../app/config-handler';

export const ConfigFieldsDefaultValues: Partial<IConfig> = {
  isPodUrlEditable: true,
  forceAutoUpdate: false,
  enableBrowserLogin: false,
  browserLoginAutoConnect: false,
  latestAutoUpdateChannelEnabled: true,
  betaAutoUpdateChannelEnabled: true,
  browserLoginRetryTimeout: '5',
  openfin: {
    uuid: '',
    licenseKey: '',
    runtimeVersion: '',
    platformUuid: '',
    autoConnect: false,
    connectionTimeout: '10000',
  },
};
