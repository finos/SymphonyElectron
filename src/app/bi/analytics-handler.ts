import { app, WebContents } from 'electron';
import * as fs from 'fs';
import * as path from 'path';
import { logger } from '../../common/logger';

export interface IAnalyticsData {
  element: AnalyticsElements;
  action_type?:
    | MenuActionTypes
    | ScreenSnippetActionTypes
    | ToastNotificationActionTypes
    | SDAUserSessionActionTypes
    | InstallActionTypes;
  action_result?: AnalyticsActions;
  extra_data?: object;
}

export interface ICrashData extends IAnalyticsData {
  process: SDACrashProcess;
  crashCause: string;
  windowName: string;
  miniDump?: string;
}

export interface ISessionData extends IAnalyticsData {
  extra_data?: {
    sessionStartDatetime?: string;
    machineStartDatetime?: string;
    machineId?: string;
    InstallVariant?: string;
    osName?: string;
    osVersion?: string;
    osLanguage?: string;
    osTimeZone?: string;
    cpuNumberOfCores?: number;
    cpuMaxFrequency?: number;
    cpuUsagePercent?: number;
    maxCPUUsagePercent?: number;
    memoryTotal?: number;
    memoryUsedPercent?: number;
    maxMemoryUsedPercent?: number;
    sdaUsedMemory?: number;
    memoryAvailable?: number;
    vdi?: boolean;
    endReason?: string;
    crashProcess?: string;
  };
}

export interface IInstallData extends IAnalyticsData {
  extra_data?: {
    installLocation: string;
    installType: string;
  };
}

export enum MenuActionTypes {
  AUTO_LAUNCH_ON_START_UP = 'auto_launch_on_start_up',
  ALWAYS_ON_TOP = 'always_on_top',
  MINIMIZE_ON_CLOSE = 'minimize_on_close',
  FLASH_NOTIFICATION_IN_TASK_BAR = 'flash_notification_in_task_bar',
  HAMBURGER_MENU = 'hamburger_menu',
  REFRESH_APP_IN_IDLE = 'refresh_app_in_idle',
}

export enum ScreenSnippetActionTypes {
  SCREENSHOT_TAKEN = 'screenshot_taken',
  ANNOTATE_ADDED_PEN = 'annotate_added_pen',
  ANNOTATE_ADDED_HIGHLIGHT = 'annotate_added_highlight',
  ANNOTATE_ADD = 'annotate_done',
  ANNOTATE_CLEARED = 'annotate_cleared',
  ANNOTATE_ERASED = 'annotate_erased',
  ANNOTATE_COPY = 'annotate_copy',
  ANNOTATE_SAVE_AS = 'annotate_save_as',
  ANNOTATE_CLOSE = 'annotate_close',
}

export enum ToastNotificationActionTypes {
  TOAST_CLOSED = 'toast_closed',
}

export enum SDAUserSessionActionTypes {
  Start = 'Start',
  End = 'End',
  Login = 'Login',
  Logout = 'Logout',
  Crash = 'Crash',
  ForceReload = 'Force_reload',
}

export enum InstallActionTypes {
  InstallStarted = 'Install_started',
  InstallCompleted = 'Install_completed',
  InstallFailed = 'Install_failed',
}

export enum InstallTypes {
  Auto = 'auto',
  Manual = 'manual',
}

export enum InstallLocationTypes {
  PROG_FILES = 'PROG_FILES',
  REMOTE = 'REMOTE',
  LOCAL = 'LOCAL',
  CUSTOM = 'CUSTOM',
}

export enum SDAEndReasonTypes {
  Reboot = 'Reboot',
  Closed = 'Closed',
  Crashed = 'Crashed',
}

export enum AnalyticsActions {
  ENABLED = 'ON',
  DISABLED = 'OFF',
}

export enum AnalyticsElements {
  MENU = 'Menu',
  SCREEN_CAPTURE_ANNOTATE = 'screen_capture_annotate',
  TOAST_NOTIFICATION = 'toast_notification',
  SDA_CRASH = 'sda_crash',
  SDA_SESSION = 'sda_session',
  SDA_INSTALL = 'sda_install',
}

export enum SDACrashProcess {
  MAIN = 'main',
  RENDERER = 'renderer',
  GPU = 'gpu',
}

const MAX_EVENT_QUEUE_LENGTH = 50;
const analyticsCallback = 'analytics-callback';

class Analytics {
  private preloadWindow: WebContents | undefined;
  private analyticsEventQueue: IAnalyticsData[] = [];
  private analyticsEndEventQueue: IAnalyticsData[] = [];
  private analyticEventsDataFilePath = path.join(
    app.getPath('userData'),
    'analytics.json',
  );

  /**
   * Stores the reference to the preload window
   *
   * @param webContents {WeContents}
   */
  public registerPreloadWindow(webContents: WebContents): void {
    this.preloadWindow = webContents;

    if (!(this.preloadWindow && !this.preloadWindow.isDestroyed())) {
      return;
    }
    if (this.analyticsEventQueue && this.analyticsEventQueue.length > 0) {
      this.analyticsEventQueue.forEach((eventData) => {
        if (this.preloadWindow && !this.preloadWindow.isDestroyed()) {
          if (eventData.element === AnalyticsElements.SDA_CRASH) {
            eventData = eventData as ICrashData;
          }
          this.preloadWindow.send(analyticsCallback, eventData);
        }
      });
      this.resetAnalytics();
    }
  }

  /**
   * Sends the analytics events to the web client
   *
   * @param eventData {IAnalyticsData}
   */
  public track(eventData: IAnalyticsData): void {
    if (eventData.element === AnalyticsElements.SDA_CRASH) {
      eventData = eventData as ICrashData;
    }
    // Store these events and send them in the next SDA session
    if (
      eventData.action_type === SDAUserSessionActionTypes.End ||
      eventData.action_type === SDAUserSessionActionTypes.Logout ||
      eventData.action_type === InstallActionTypes.InstallStarted
    ) {
      this.analyticsEndEventQueue.push(eventData);
      return;
    }
    this.sendAnalyticsOrAddToQueue(eventData);
  }

  /**
   * Writes all the pending stats into a file
   */
  public writeAnalyticFile = () => {
    try {
      fs.writeFileSync(
        this.analyticEventsDataFilePath,
        JSON.stringify(this.analyticsEndEventQueue, null, 2),
        { encoding: 'utf8' },
      );
      logger.info(
        `analytics-handler: updated analytic values with the data ${JSON.stringify(
          this.analyticsEndEventQueue,
        )}`,
      );
    } catch (error) {
      logger.error(
        `analytics-handler: failed to update analytic with ${JSON.stringify(
          this.analyticsEndEventQueue,
        )}`,
        error,
      );
    }
  };

  /**
   * Sends all the locally stored stats
   */
  public sendLocalAnalytics = () => {
    if (fs.existsSync(this.analyticEventsDataFilePath)) {
      const localStats = fs.readFileSync(
        this.analyticEventsDataFilePath,
        'utf8',
      );
      if (!localStats) {
        return;
      }
      let parsedStats: ISessionData[];
      try {
        parsedStats = JSON.parse(localStats);
        logger.info(
          `analytics-handler: parsed stats JSON file with data`,
          parsedStats,
        );
        if (parsedStats && parsedStats.length) {
          parsedStats.forEach((event) => {
            this.sendAnalyticsOrAddToQueue(event);
          });
          fs.unlinkSync(this.analyticEventsDataFilePath);
        }
      } catch (e: any) {
        logger.error(
          `analytics-handler: parsing stats JSON file failed due to error ${e}`,
        );
      }
    }
    if (this.analyticsEndEventQueue.length > 0) {
      this.analyticsEndEventQueue.forEach((eventData) => {
        this.sendAnalyticsOrAddToQueue(eventData);
      });
      this.analyticsEndEventQueue = [];
    }
  };

  /**
   * Clears the analytics queue
   */
  public resetAnalytics(): void {
    this.analyticsEventQueue = [];
  }

  private sendAnalyticsOrAddToQueue = (eventData) => {
    if (this.preloadWindow && !this.preloadWindow.isDestroyed()) {
      this.preloadWindow.send(analyticsCallback, eventData);
      return;
    }
    this.analyticsEventQueue.push(eventData);
    // don't store more than specified limit. keep most recent events.
    if (this.analyticsEventQueue.length > MAX_EVENT_QUEUE_LENGTH) {
      this.analyticsEventQueue.shift();
    }
  };
}

const analytics = new Analytics();

export { analytics };
