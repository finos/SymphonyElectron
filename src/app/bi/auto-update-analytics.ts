import { app } from 'electron';
import { isMac, isWindowsOS } from '../../common/env';
import {
  analytics,
  AnalyticsElements,
  IInstallData,
  InstallActionTypes,
  InstallLocationTypes,
  InstallTypes,
} from './analytics-handler';

/**
 * Sends auto update analytics event
 * @param action
 * @param installType
 */
export const sendAutoUpdateAnalytics = (
  action: InstallActionTypes,
  installType: InstallTypes,
) => {
  const installLocation = getInstallLocation();
  const event: IInstallData = {
    element: AnalyticsElements.SDA_INSTALL,
    action_type: action,
    extra_data: {
      installLocation,
      installType,
    },
  };
  analytics.track(event);
  if (action === InstallActionTypes.InstallStarted) {
    analytics.writeAnalyticFile();
  }
};

/**
 * Identifies and returns the installation location
 */
const getInstallLocation = () => {
  const appPath = app.getPath('exe');
  if (isWindowsOS) {
    if (appPath.includes('AppData\\Local\\Programs')) {
      return InstallLocationTypes.LOCAL;
    }
    if (appPath.includes('Program Files')) {
      return InstallLocationTypes.PROG_FILES;
    }
    return InstallLocationTypes.CUSTOM;
  }
  if (isMac) {
    if (appPath.includes('/Applications')) {
      return InstallLocationTypes.PROG_FILES;
    }
    return InstallLocationTypes.LOCAL;
  }

  return InstallLocationTypes.PROG_FILES;
};
