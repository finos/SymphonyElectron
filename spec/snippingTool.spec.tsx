jest.mock('save-svg-as-png', function () {
  return {
    svgAsPngUri: async function (svg) {
      return Promise.resolve(svg);
    },
  };
});

import { mount, shallow } from 'enzyme';
import * as React from 'react';
import { ScreenSnippetActionTypes } from '../src/app/bi/analytics-handler';
import { ScreenShotAnnotation } from '../src/common/ipcEvent';
import SnippingTool from '../src/renderer/components/snipping-tool';
import { ipcRenderer } from './__mocks__/electron';

const waitForPromisesToResolve = () =>
  new Promise((resolve) => setTimeout(resolve));

describe('Snipping Tool', () => {
  afterEach(() => {
    jest.clearAllMocks();
  });

  it('should render correctly', () => {
    const wrapper = shallow(React.createElement(SnippingTool));
    expect(wrapper).toMatchSnapshot();
  });

  it('should set up a "once" listener for snipping-tool-data event on mounting', () => {
    const spy = jest.spyOn(ipcRenderer, 'once');
    mount(React.createElement(SnippingTool));
    expect(spy).toBeCalledWith('snipping-tool-data', expect.any(Function));
  });

  it('should send screenshot_taken BI event on component mount', () => {
    const spy = jest.spyOn(ipcRenderer, 'send');
    const expectedValue = {
      type: 'screenshot_taken',
      element: 'screen_capture_annotate',
    };
    mount(React.createElement(SnippingTool));
    expect(spy).toBeCalledWith('snippet-analytics-data', expectedValue);
  });

  it('should send capture_sent BI event when clicking done', async () => {
    const spy = jest.spyOn(ipcRenderer, 'send');
    const expectedValue = {
      type: ScreenSnippetActionTypes.ANNOTATE_ADD,
      element: 'screen_capture_annotate',
    };
    const wrapper = mount(React.createElement(SnippingTool));
    wrapper.find('[data-testid="done-button"]').simulate('click');
    wrapper.update();
    await waitForPromisesToResolve();
    expect(spy).toBeCalledWith('snippet-analytics-data', expectedValue);
  });

  it('should send capture_sent BI event when clicking copy to clipboard', async () => {
    const spy = jest.spyOn(ipcRenderer, 'send');
    const expectedValue = {
      type: ScreenSnippetActionTypes.ANNOTATE_COPY,
      element: 'screen_capture_annotate',
    };
    const wrapper = mount(React.createElement(SnippingTool));
    wrapper.find('[data-testid="snipping-tool_MENU_BUTTON"]').simulate('click');
    wrapper.update();
    await waitForPromisesToResolve();
    wrapper
      .find('[data-testid="snipping-tool_COPY_TO_CLIPBOARD"]')
      .simulate('click');
    wrapper.update();
    await waitForPromisesToResolve();
    expect(spy).toBeCalledWith('snippet-analytics-data', expectedValue);
  });

  it('should send capture_sent BI event when clicking save as', async () => {
    const spy = jest.spyOn(ipcRenderer, 'send');
    const expectedValue = {
      type: ScreenSnippetActionTypes.ANNOTATE_SAVE_AS,
      element: 'screen_capture_annotate',
    };
    const wrapper = mount(React.createElement(SnippingTool));
    wrapper.find('[data-testid="snipping-tool_MENU_BUTTON"]').simulate('click');
    wrapper.update();
    await waitForPromisesToResolve();
    wrapper.find('[data-testid="snipping-tool_SAVE_AS"]').simulate('click');
    wrapper.update();
    await waitForPromisesToResolve();
    expect(spy).toBeCalledWith('snippet-analytics-data', expectedValue);
  });

  it('should send capture_sent BI event when clicking done', async () => {
    const spy = jest.spyOn(ipcRenderer, 'send');
    const expectedValue = {
      type: ScreenSnippetActionTypes.ANNOTATE_CLOSE,
      element: 'screen_capture_annotate',
    };
    const wrapper = mount(React.createElement(SnippingTool));
    wrapper.find('[data-testid="close-button"]').simulate('click');
    wrapper.update();
    await waitForPromisesToResolve();
    expect(spy).toBeCalledWith('snippet-analytics-data', expectedValue);
  });

  it('should send annotate_cleared BI event when clicking clear', async () => {
    const spy = jest.spyOn(ipcRenderer, 'send');
    const expectedValue = {
      type: 'annotate_cleared',
      element: 'screen_capture_annotate',
    };
    const wrapper = mount(React.createElement(SnippingTool));
    wrapper.find('[data-testid="clear-button"]').simulate('click');
    wrapper.update();
    await waitForPromisesToResolve();
    expect(spy).toBeCalledWith('snippet-analytics-data', expectedValue);
  });

  it('should render pen color picker when clicked on pen', () => {
    const wrapper = shallow(React.createElement(SnippingTool));
    wrapper.find('[data-testid="pen-button"]').simulate('click');
    expect(wrapper.find('[data-testid="pen-colorpicker"]').exists()).toBe(true);
  });

  it('should render highlight color picker when clicked on highlight', () => {
    const wrapper = shallow(React.createElement(SnippingTool));
    wrapper.find('[data-testid="highlight-button"]').simulate('click');
    expect(wrapper.find('[data-testid="highlight-colorpicker"]').exists()).toBe(
      true,
    );
  });

  it('should clear all paths when clicked on clear', () => {
    const props = {
      existingPaths: [
        {
          points: [{ x: 0, y: 0 }],
          shouldShow: true,
          strokeWidth: 5,
          color: 'rgba(233, 0, 0, 0.64)',
          key: 'path0',
        },
      ],
    };
    const wrapper = mount(<SnippingTool {...props} />);
    const annotateComponent = wrapper.find(
      '[data-testid="annotate-component"]',
    );
    wrapper.find('[data-testid="clear-button"]').simulate('click');
    wrapper.update();
    expect(annotateComponent.prop('paths')).toEqual([
      {
        color: 'rgba(233, 0, 0, 0.64)',
        key: 'path0',
        points: [{ x: 0, y: 0 }],
        shouldShow: false,
        strokeWidth: 5,
      },
    ]);
  });

  it('should send upload-snippet event with correct data when clicked on done', async () => {
    const spy = jest.spyOn(ipcRenderer, 'send');
    const wrapper = mount(<SnippingTool />);
    wrapper.find('[data-testid="done-button"]').simulate('click');
    wrapper.update();
    await waitForPromisesToResolve();
    expect(spy).toBeCalledWith('upload-snippet', {
      mergedImageData: 'MERGE_FAIL',
      screenSnippetPath: '',
    });
  });

  it('should send upload-snippet event with correct data when clicked on copy to clipboard', async () => {
    const wrapper = mount(<SnippingTool />);
    const spy = jest.spyOn(ipcRenderer, 'send');
    jest.spyOn(document, 'getElementById').mockImplementation((selector) => {
      switch (selector) {
        case 'annotate-area':
          return '123';
        default:
          return '';
      }
    });

    wrapper.find('[data-testid="snipping-tool_MENU_BUTTON"]').simulate('click');
    wrapper.update();
    await waitForPromisesToResolve();
    wrapper
      .find('[data-testid="snipping-tool_COPY_TO_CLIPBOARD"]')
      .simulate('click');

    wrapper.update();
    await waitForPromisesToResolve();
    expect(spy).toBeCalledWith(ScreenShotAnnotation.COPY_TO_CLIPBOARD, {
      clipboard: '123',
    });
  });

  it('should send upload-snippet event with correct data when clicked on save as', async () => {
    const wrapper = mount(<SnippingTool />);
    const spy = jest.spyOn(ipcRenderer, 'send');
    jest.spyOn(document, 'getElementById').mockImplementation((selector) => {
      switch (selector) {
        case 'annotate-area':
          return '123';
        default:
          return '';
      }
    });
    wrapper.find('[data-testid="snipping-tool_MENU_BUTTON"]').simulate('click');
    wrapper.update();
    await waitForPromisesToResolve();
    wrapper.find('[data-testid="snipping-tool_SAVE_AS"]').simulate('click');
    wrapper.update();
    await waitForPromisesToResolve();
    expect(spy).toBeCalledWith(ScreenShotAnnotation.SAVE_AS, {
      clipboard: '123',
    });
  });

  it('should send upload-snippet event with correct data when clicked on close', async () => {
    const wrapper = mount(<SnippingTool />);
    const spy = jest.spyOn(ipcRenderer, 'send');
    wrapper.find('[data-testid="close-button"]').simulate('click');
    wrapper.update();
    await waitForPromisesToResolve();
    expect(spy).toBeCalledWith(ScreenShotAnnotation.CLOSE);
  });
});
