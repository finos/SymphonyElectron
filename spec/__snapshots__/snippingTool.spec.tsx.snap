// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snipping Tool should render correctly 1`] = `
<div
  className="snipping-tool"
  lang="en-US"
>
  <header>
    <div
      className="draw-actions"
    >
      <button
        className="action-button"
        data-testid="pen-button"
        onClick={[Function]}
        style={
          Object {
            "border": "2px solid rgba(0, 142, 255, 1)",
          }
        }
        title="Pen"
      >
        <img
          draggable={false}
          src={Object {}}
        />
      </button>
      <button
        className="action-button"
        data-testid="highlight-button"
        onClick={[Function]}
        title="Highlight"
      >
        <img
          draggable={false}
          src={Object {}}
        />
      </button>
      <button
        className="action-button"
        data-testid="erase-button"
        onClick={[Function]}
        title="Erase"
      >
        <img
          draggable={false}
          src={Object {}}
        />
      </button>
    </div>
    <div
      className="clear-actions"
    >
      <button
        className="clear-button"
        data-testid="clear-button"
        onClick={[Function]}
      >
        Clear
      </button>
    </div>
  </header>
  <main>
    <div
      className="image-container"
    >
      <AnnotateArea
        annotateAreaDimensions={
          Object {
            "height": 600,
            "width": 800,
          }
        }
        backgroundImagePath=""
        chosenTool="PEN"
        data-testid="annotate-component"
        highlightColor="rgba(0,  142,  255, 0.3"
        imageDimensions={
          Object {
            "height": 600,
            "width": 800,
          }
        }
        onChange={[Function]}
        paths={Array []}
        penColor="rgba(0, 142, 255, 1)"
      />
    </div>
  </main>
  <footer>
    <button
      className="close-button"
      data-testid="close-button"
      onClick={[Function]}
    >
      Close
    </button>
    <button
      className="add-to-chat-button"
      data-testid="done-button"
      onClick={[Function]}
    >
      Add to chat
    </button>
    <MenuButton
      id="snipping-tool"
      listItems={
        Array [
          Object {
            "dataTestId": "COPY_TO_CLIPBOARD",
            "event": "copy-to-clipboard",
            "name": "Copy to clipboard",
            "onClick": [Function],
          },
          Object {
            "dataTestId": "SAVE_AS",
            "event": "save-as",
            "name": "Save as",
            "onClick": [Function],
          },
        ]
      }
    />
  </footer>
</div>
`;
